src/samplers/adapt/precond.jl

```diff
@@ -1,43 +1,85 @@
 ##########################
 ### Variance estimator ###
 ##########################
+abstract type VarEstimator{TI,TF} end
+
 # Ref： https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_var_estimator.hpp
-mutable struct VarEstimator{TI<:Integer,TF<:Real}
+mutable struct WelfordVar{TI,TF} <: VarEstimator{TI,TF}
     n :: TI
     μ :: Vector{TF}
     M :: Vector{TF}
 end
 
-function reset!(ve::VarEstimator{TI,TF}) where {TI<:Integer,TF<:Real}
-    ve.n = zero(TI)
-    ve.μ .= zero(TF)
-    ve.M .= zero(TF)
+function reset!(wv::WelfordVar{TI,TF}) where {TI<:Integer,TF<:Real}
+    wv.n = zero(TI)
+    wv.μ .= zero(TF)
+    wv.M .= zero(TF)
 end
 
-function add_sample!(ve::VarEstimator, s::AbstractVector)
-    ve.n += 1
-    δ = s .- ve.μ
-    ve.μ .+= δ ./ ve.n
-    ve.M .+= δ .* (s .- ve.μ)
+function add_sample!(wv::WelfordVar, s::AbstractVector)
+    wv.n += 1
+    δ = s .- wv.μ
+    wv.μ .+= δ ./ wv.n
+    wv.M .+= δ .* (s .- wv.μ)
 end
 
 # https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/var_adaptation.hpp
-function get_var(ve::VarEstimator)
-    n, M = ve.n, ve.M
+function get_var(wv::VarEstimator{TI,TF})::Vector{TF} where {TI<:Integer,TF<:Real}
+    n, M = wv.n, wv.M
     @assert n >= 2 "Cannot get variance with only one sample"
-    return (n / ((n + 5) * (n - 1))) .* M .+ 1e-3 * (5.0 / (n + 5))
+    return (n / ((n + 5) * (n - 1))) .* M .+ 1e-3 * (5 / (n + 5))
+end
+
+abstract type CovarEstimator{TI<:Integer,TF<:Real} end
```

Comment:
Do we need two abstract types (VarEstimator, CovarEstimator)? Perhaps consider merge them into one abstract type.

Reply:
We currently have both naive covariance computation and the Welford one. The DensePreConditioner can use either of them